File: kmp/ver-nodejs/src/commonMain/kotlin/org/opengamestudio/srv.kt
package org.opengamestudio
import kotlin.js.JsExport

//<!-- Constants -->

@JsExport val SRV_API_ROOT = "/"

@JsExport val SRV_ARGUMENT_BROWSER_DIR = "--browserDir"
@JsExport val SRV_ARGUMENT_PROJECT_DIR = "--projectDir"

@JsExport val SRV_DEFAULT_HTTP_PORT = 8000
@JsExport val SRV_INDEX = "index.html"

//<!-- Shoulds -->
 
/* List directory files
 *
 * Conditions:
 * 1. POST /list
 */
@JsExport
fun srvShouldListDir(c: SrvContext): SrvContext {
    if (
        c.recentField == "request" &&
        c.request.method == CONST_POST && 
        c.request.url == CONST_API_LIST
    ) {
        c.listDir = "${c.projectAbsPath}/${c.request.body}"
        c.recentField = "listDir"
        return c
    }


    c.recentField = "none"
    return c
}

/* Open URL in browser
 *
 * Conditions:
 * 1. Did launch
 */
@JsExport
fun srvShouldOpenURL(c: SrvContext): SrvContext {
    if (c.recentField == "didLaunch") {
        c.url = "http://localhost:" + c.httpPort
        c.recentField = "url"
        return c
    }

    c.recentField = "none"
    return c
}

/* Read a file
 *
 * Conditions:
 * 1. GET /
 * 2. GET /<file> (excluding reserved API calls)
 * 3. POST /read
 */
@JsExport
fun srvShouldReadFile(c: SrvContext): SrvContext {
    if (
        c.recentField == "request" &&
        c.request.method == CONST_GET && 
        c.request.url == SRV_API_ROOT
    ) {
        c.readFile = "${c.browserDir}/${SRV_INDEX}"
        c.recentField = "readFile"
        return c
    }

    if (
        c.recentField == "request" &&
        c.request.method == CONST_GET && 
        c.request.url != CONST_API_PATH
    ) {
        c.readFile = "${c.browserDir}${c.request.url}"
        c.recentField = "readFile"
        return c
    }

    if (
        c.recentField == "request" &&
        c.request.method == CONST_POST && 
        c.request.url == CONST_API_READ
    ) {
        c.readFile = "${c.projectAbsPath}/${c.request.body}"
        c.recentField = "readFile"
        return c
    }


    c.recentField = "none"
    return c
}

/* Reset path to serve files requested by web browser
 *
 * Conditions:
 * 1. Did launch
 */
@JsExport
fun srvShouldResetBrowserDir(c: SrvContext): SrvContext {
    if (c.recentField == "didLaunch") {
        c.browserDir = cliArgumentValue(c.arguments, SRV_ARGUMENT_BROWSER_DIR)
        c.recentField = "browserDir"
        return c
    }

    c.recentField = "none"
    return c
}

/* HTTP port to use
 *
 * Conditions:
 * 1. Did set default HTTP port
 */
@JsExport
fun srvShouldResetHTTPPort(c: SrvContext): SrvContext {
    if (c.recentField == "defaultHTTPPort") {
        c.httpPort = c.defaultHTTPPort
        c.recentField = "httpPort"
        return c
    }

    c.recentField = "none"
    return c
}

/* Reset path to where PSKOV files are located
 *
 * Conditions:
 * 1. Did launch
 */
@JsExport
fun srvShouldResetProjectDir(c: SrvContext): SrvContext {
    if (c.recentField == "didLaunch") {
        c.projectDir = cliArgumentValue(c.arguments, SRV_ARGUMENT_PROJECT_DIR)
        c.recentField = "projectDir"
        return c
    }

    c.recentField = "none"
    return c
}

/* Generate HTTP response
 *
 * Conditions:
 * 1. Did receive contents of the requested file
 * 2. GET /path
 * 3. Did receive contents of the requested directory
 * 4. Did finish writing a file
 */
@JsExport
fun srvShouldResetResponse(c: SrvContext): SrvContext {
    if (c.recentField == "readFileContents") {
        c.response = NetResponse(c.readFileContents, c.request)
        c.recentField = "response"
        return c
    }

    if (
        c.recentField == "request" &&
        c.request.method == CONST_GET &&
        c.request.url == CONST_API_PATH
    ) {
        c.response = NetResponse(c.projectAbsPath, c.request)
        c.recentField = "response"
        return c
    }

    if (c.recentField == "dirFiles") {
        val json = filesToJSON(c.dirFiles)
        c.response = NetResponse(json, c.request)
        c.recentField = "response"
        return c
    }

    if (c.recentField == "didWriteFile") {
        var reply = ""
        if (!c.didWriteFile) {
            reply = "ERR write"
        }
        c.response = NetResponse(reply, c.request)
        c.recentField = "response"
        return c
    }

    c.recentField = "none"
    return c
}

/* Write a file
 *
 * Conditions:
 * 1. POST /write
 */
@JsExport
fun srvShouldWriteFile(c: SrvContext): SrvContext {
    if (
        c.recentField == "request" &&
        c.request.method == CONST_POST && 
        c.request.url == CONST_API_WRITE
    ) {
        val d = jsonToFileContents(c.request.body)
        val fullPath = "${c.projectAbsPath}/${d["path"]!!}"
        c.writeFile = arrayOf(fullPath, d["contents"]!!)
        c.recentField = "writeFile"
        return c
    }

    c.recentField = "none"
    return c
}

File: kmp/ver-nodejs/src/jsMain/resources/srv.js
#!/usr/bin/env node

let fs = require("fs");
let http = require("http");
let mime = require("mime-types");
let open = require("open");
let path = require("path");

let other = require("./other");
let KT = require("./pskov-ver-nodejs").org.opengamestudio;

//!<-- API -->

function srvCtrl() {
    return cmp.ctrl;
}

//!<-- Constants -->

let SRV_CONTENTS_404 = "404";
let SRV_CONTENTS_BINARY = "BINARY";

//!<-- Component -->

function SrvComponent() {
    this._construct = function() {
        this.ctrl = new KT.CLDController(new KT.SrvContext());
        other.registerCtrlDbgOutput(this.ctrl, "Srv", KT);

        this.setupEffects();
        this.setupShoulds();
    };

    this.setupEffects = function() {
        let oneliners = [ 
            "deleteFile", (c) => { srvDeleteFile(c.deleteFile) },
            "listDir", (c) => { srvListDir(c.listDir) },
            "projectDir", (c) => { srvResolvePath(c.projectDir) },
            "readFile", (c) => { srvReadFile(c.readFile) },
            "url", (c) => { open(c.url) },
            "writeFile", (c) => { srvWriteFile(c.writeFile[0], c.writeFile[1]) },
        ];
        KT.registerOneliners(this.ctrl, oneliners);
    };

    this.setupShoulds = function() {
        [
            KT.srvShouldDeleteFile,
            KT.srvShouldListDir,
            KT.srvShouldOpenURL,
            KT.srvShouldReadFile,
            KT.srvShouldResetBrowserDir,
            KT.srvShouldResetHTTPPort,
            KT.srvShouldResetProjectDir,
            KT.srvShouldResetResponse,
            KT.srvShouldWriteFile,
        ].forEach((f) => {
            this.ctrl.registerFunction(f);
        });
    };

    this._construct();
}

//<!-- Effects -->

function srvDeleteFile(fileName) {
    var isOk = true;
    try {
        fs.rmSync(fileName);
    } catch (e) {
        console.error("ERR srvDF e:", e);
        isOk = false;
    }
    srvCtrl().set("didDeleteFile", isOk);
}

function srvListDir(path) {
    var files = [];

    try {
        var names = fs.readdirSync(path);
        names.sort();
        for (let i in names) {
            let name = names[i];
            let filePath = path + "/" + name;
            let isFile = fs.statSync(filePath).isFile();
            let f = new KT.FSFile(isFile, name);
            files.push(f);
        }
    } catch (e) {
        console.log(e);
    }

    srvCtrl().set("dirFiles", files);
}

function srvReadFile(fileName) {
    var contents = "";
    let type = mime.lookup(fileName);
    let isText = srvIsTextFile(type, fileName);

    // Binary
    if (!isText) {
        contents = SRV_CONTENTS_BINARY;
    }

    // Text
    if (isText) {
        try {
            contents = fs.readFileSync(fileName, { encoding: "utf8", flag: "r" });
        } catch (e) {
            contents = SRV_CONTENTS_404;
        }
    }
    srvCtrl().set("readFileContents", contents);
}

function srvResolvePath(relative) {
    let p = path.resolve(relative);
    srvCtrl().set("projectAbsPath", p);
}

function srvWriteFile(fileName, contents) {
    var isOk = true;
    try {
        fs.writeFileSync(fileName, contents);
    } catch (e) {
        console.error("ERR srvWF e:", e);
        isOk = false;
    }
    srvCtrl().set("didWriteFile", isOk);
}

//<!-- Functions -->

// Detect if served file is text
//
// Conditions:
// 1. MIME type was not detected -> Text
// 2. Specified MIME type -> Text
// 3. Unspecified MIME type -> Binary
function srvIsTextFile(mimeType, fileName) {
    /* 1 */ if (!mimeType) {
        return true;
    }

    /* 2 */ if (
        mimeType.startsWith("text") ||
        mimeType.includes("svg")
    ) {
        return true;
    }

    /* 3 */ return false;
}

//<!-- Installation -->

let cmp = new SrvComponent();

//<!-- Defaults -->

srvCtrl().set("arguments", process.argv);
srvCtrl().set("defaultBrowserDir", __dirname + "/../ver-browser");
srvCtrl().set("defaultHTTPPort", KT.SRV_DEFAULT_HTTP_PORT);

//<!-- Server -->

let srv = http.createServer((req, res) => {
    // Collect request body
    var body = "";
    req.on("data", (chunk) => {
        body = chunk.toString();
    });

    // Process request when finished collecting the body
    req.on("end", () => {
        let netRequest = new KT.NetRequest(body, req.method, req.url);
        srvCtrl().set("request", netRequest);
        let response = srvCtrl().context.response;
 
        // Allow CORS
        res.setHeader("Access-Control-Allow-Origin", "*");

        // File does not exist
        if (response.contents == SRV_CONTENTS_404) {
            res.writeHead(404);
            res.end();
            return;
        }

        let type = mime.lookup(req.url);
        let isText = srvIsTextFile(type, req.url);

        // Binary file
        if (response.contents == SRV_CONTENTS_BINARY) {
            res.setHeader("Content-Type", type);
            try {
                let fileName = srvCtrl().context.readFile;
                let buf = fs.readFileSync(fileName, { flag: "r" });
                res.writeHead(200);
                res.end(buf, "binary");
            } catch (e) {
                res.writeHead(404);
                res.end();
            }
            return;
        }

        // Text file
        res.setHeader("Content-Type", type);
        res.writeHead(200);
        res.end(response.contents);
    });
});

srv.listen(srvCtrl().context.httpPort, (e) => {
  srvCtrl().set("didLaunch", true);
});
